### PR TITLE
codec/raw: implement the raw codec

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
I also could not find a spec doc for it, otherwise I would have linked it.

File: codec/raw/codec.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+// TODO(mvdan): make go-ipld-prime use go-multicodec soon
+const rawMulticodec = 0x55
+
+func init() {
+	cidlink.RegisterMulticodecDecoder(rawMulticodec, Decoder)
+	cidlink.RegisterMulticodecEncoder(rawMulticodec, Encoder)
+}
+
+type byteAccesor interface {
```
Comment (warpfork) [inline codec/raw/codec.go:31]: Okay, I know this is copying a typo elsewhere, but `s/Accesor/Accessor/g` please

File: codec/raw/codec.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	Bytes() []byte
+}
+
+func Decoder(am ipld.NodeAssembler, r io.Reader) error {
```
Comment (warpfork) [inline codec/raw/codec.go:35]: Would prefer to call this `Decode` (minus the 'r') now.  (I've changed most of the other functions to this grammar in #143.)

File: codec/raw/codec.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	// We need r's bytes, and r is often just a buffer.
+	// So if we can get them directly without ReadAll's copying,
+	// just do that and also save an allocation.
+	// Nodes are immutable, so we won't modify the bytes.
+	// We assume that the incoming buffer won't get modified either.
```
Comment (warpfork) [inline codec/raw/codec.go:40]: I think we should move this to the actual docs block on the function.  (I'm probably guilty of putting too many tidbits like this in comments too, but I should also quit it.)

I'm trying to get in the habit of using phrasing about "it is not lawful to xyz" ([example](https://github.com/ipld/go-ipld-prime/blob/7d918468f40d06d8a64a2fb6c162134da497bebb/path.go#L152)) for things like this mutability contract that we can't really enforce with the compiler.  Just to have _some_ convention for talking about this semi-formally.

**Review by warpfork — APPROVED** [review]
I had some nits to pick, but this is holistically pretty good.

File: codec/raw/codec.go (hunk 1 above)
Comment (mvdan) [reply codec/raw/codec.go:31]: actually, there's no need for this to be a named type. I'll just use an anonymous interface.

File: codec/raw/codec.go (hunk 2 above)
Comment (mvdan) [reply codec/raw/codec.go:35]: Ah, right, thanks for reminding me.

File: codec/raw/codec.go (hunk 3 above)
Comment (mvdan) [reply codec/raw/codec.go:40]: Fair enough. I thought about it initially, and thought that some of these details are too internal for the godoc. But I think you're right, it's still better than no godoc.

**Comment by mvdan** [comment]
@warpfork ready, I think.

File: codec/raw/codec.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// Note that if r has a Bytes method, such as is the case with *bytes.Buffer, we
+// will use those bytes directly to save having to allocate and copy them. The
+// Node interface is defined as immutable, so it is assumed that its bytes won't
+// be modified in-place. Similarly, we assume that the incoming buffer's bytes
+// won't get modified in-place later.
+//
+// To disable the shortcut above, hide the Bytes method by wrapping the buffer
+// with an io.Reader:
+//
+//     Decode([...], struct{io.Reader}{buf})
```
Comment (warpfork) [inline codec/raw/codec.go:37]: oh wow, TIL.  That's a neat idiom.

**Review by warpfork — APPROVED** [review]
:ship:

File: codec/raw/codec.go (hunk 4 above)
Comment (mvdan) [reply codec/raw/codec.go:37]: one can accomplish it the explicit way, like I did in the test, but this shortcut is much shorter. It's not easy to realise it's possible unless you've seen it before, so... doc :)

**Pull request merged**